src/org/benf/cfr/reader/bytecode/analysis/parse/rewriters/ConstantFoldingRewriter.java

```diff
@@ -0,0 +1,54 @@
+package org.benf.cfr.reader.bytecode.analysis.parse.rewriters;
+
+import org.benf.cfr.reader.bytecode.analysis.parse.Expression;
+import org.benf.cfr.reader.bytecode.analysis.parse.LValue;
+import org.benf.cfr.reader.bytecode.analysis.parse.StatementContainer;
+import org.benf.cfr.reader.bytecode.analysis.parse.expression.ArithmeticMonOperation;
+import org.benf.cfr.reader.bytecode.analysis.parse.expression.ArithmeticOperation;
+import org.benf.cfr.reader.bytecode.analysis.parse.expression.CastExpression;
+import org.benf.cfr.reader.bytecode.analysis.parse.expression.Literal;
+import org.benf.cfr.reader.bytecode.analysis.parse.utils.SSAIdentifiers;
+import org.benf.cfr.reader.bytecode.analysis.types.RawJavaType;
+import org.benf.cfr.reader.util.collections.MapFactory;
+
+import java.util.Map;
+
+public class ConstantFoldingRewriter extends AbstractExpressionRewriter {
+	public static final ConstantFoldingRewriter INSTANCE = new ConstantFoldingRewriter();
+	private static final Map<LValue, Literal> DISPLAY_MAP = MapFactory.<LValue, Literal>newMap();
+
+	@Override
+	public Expression rewriteExpression(Expression expression, SSAIdentifiers ssaIdentifiers, StatementContainer statementContainer, ExpressionRewriterFlags flags) {
+		expression.applyExpressionRewriter(this, ssaIdentifiers, statementContainer, flags);
+		// Skip if expression type is non-primitive
+		if (expression.getInferredJavaType().getRawType().ordinal() > RawJavaType.DOUBLE.ordinal())
+			return expression;
+		// Simplify arithmetic
+		if (expression instanceof ArithmeticOperation || expression instanceof ArithmeticMonOperation) {
```

Comment:
Is there any particular reason for the explicit tests?  

```
			Expression computed = expression.getComputedLiteral(getDisplayMap());
			if (computed != null) {
				expression = computed;
			}
```

would work just as well, and be more general - if something doesn't support it, it will return quickly...

Reply:
No real reason. I just wanted to avoid potentially unknown behavior outside of the scope of this PR. I can skip the checks.